kube-core/src/conversion.rs

```diff
@@ -0,0 +1,141 @@
+//! Contains utilities for implements custom resource conversion webhooks.
+//! Abstractions in this module and its submodules may be ordered:
+//! - `low_level` submodule contains raw typings, allowing for full control (probably you don't need these unless you are implementing your own apiserver).
+//! - `Converter` together with `ConversionHandler` are middle ground: they abstract you from low-level details but not require particular conversion strategy.
+//! - `StarConverter` is opinionated strategy, where all conversions are done using intermediate unversioned representation as a middle point.
+//! Third option is preferred - this is how Kubernetes implements versioning and convertion for builtin APIs.
+
+use std::sync::Arc;
+
+use self::low_level::{ConversionResponse, ConversionReview};
+
+/// Defines low-level typings.
+pub mod low_level;
+/// High-level easy-to-use converter.
+mod star;
+
+pub use star::{StarConverter, StarConverterBuilder, StarRay};
+
+/// Helper which implements low-level API based on the provided converter.
+#[derive(Clone)]
+pub struct ConversionHandler {
+    converter: Arc<dyn Converter + Send + Sync>,
+}
+
+impl ConversionHandler {
+    /// Creates new `ConversionHandler` which will use given `converter` for operation.
+    pub fn new<C: Converter + Send + Sync + 'static>(converter: C) -> Self {
+        ConversionHandler {
+            converter: Arc::new(converter),
+        }
+    }
+
+    /// Processes `request` using stored converter so that you don't need
+    /// to think about copying `.uid` and other boring things.
+    /// Returned value is ready to be serialized and returned.
+    pub fn handle(&self, review: ConversionReview) -> ConversionReview {
+        let req = match review.request {
+            Some(r) => r,
+            None => {
+                return ConversionReview {
+                    types: review.types.clone(),
+                    request: None,
+                    response: Some(ConversionResponse::error(
+                        String::new(),
+                        ".request is unset in input".to_string(),
+                    )),
+                }
+            }
+        };
```

Comment:
Done. However I currently exposed this conversion as `from_review` method, rather than as `TryInto` impl. It seems unlikely that user will do this conversion in generic context, and method is better visible that trait impl.

Reply:
I think that is sensible, but I think it's reasonable to provide both in this case, some people look for From/TryFrom impls, to look for conversions, other scan all the methods. If `from_review` just reference a try_into impl then that's nice.

In fact I think it's an oversight that admission only does this for one way, and should also change it's `into_review` to just reference a `From` impl:
```rust
impl From<AdmissionResponse> for AdmissionReview<DynamicObject> {
    fn from(ar: AdmissionResponse) -> Self {
        Self {
            types: ar.types.clone(),
            request: None,
            response: Some(ar),
        }
    }
}
```